Feature 23

DmitriyYakubovskiy opened the pull request.

CozyWool committed:
Начал писать логику сервера

DmitriyYakubovskiy committed:
Начало реализации логики main серверы

DmitriyYakubovskiy committed:
Добавил регистрацию через сервер

от сервера нужен доступ к MappingUser и к моделям, предлагаю перенести в Quizzify.DataAccess

thisfozz commented on Quizzify/Quizzify.Client/ViewModel/RegistrationViewModel.cs:
```diff
@@ -93,29 +95,25 @@ public RegistrationViewModel(IConfiguration configuration)
+        if (signal.IsRegistered == null) MessageBox.Show("Превышено время ожидания");
+        else if (signal.IsRegistered == true) MessageBox.Show("Пользователь зарегистрирован!");
+        else MessageBox.Show("Ошибка");
     }
```
Может все таки оставить заполнение newUser  на клиенте, а потом уже в конце делать отправку на сервер

`private void RegistrationUser(object obj)
    {
        var aes = new AESManager();
        Guid guid = Guid.NewGuid();

        byte[] saltBytes = guid.ToByteArray();
        string salt = Convert.ToBase64String(saltBytes);

        string encryptedPassword = aes.Encrypt(UserPassword, salt);

        var newUser = new RegistrationModel()
        {
            UserId = guid,
            Login = UserLogin,
            Password = encryptedPassword,
            Email = UserEmail,
            SelectedSecretQuestionId = UserSelectedSecretQuestionId,
            SecretAnswer = UserSecretAnswer
        };

        // Отправляем  newUser на сервер
    }
`

thisfozz commented on Quizzify/Quizzify.MainServer/Hubs/MainHub.cs:
```diff
@@ -0,0 +1,62 @@
+            return true;
+        }
+
+        private bool RegistrationVerification(string userLogin, string userPassword, string userEmail, int userSelectedSecretQuestionId, string userSecretAnswer)
```
А тут ждать newUser, который добавляешь через контекст

DmitriyYakubovskiy committed:
Добавил регистрацию

thisfozz commented on Quizzify/Quizzify.Client/ViewModel/AuthorizationViewModel.cs:
```diff
@@ -1,6 +1,66 @@
+        }
+    }
+
+    private async void RegistrationUser(object obj)
```
Разве метод "RegistrationUser" должен называться в авторизации?

thisfozz commented on Quizzify/Quizzify.MainServer/Hubs/MainHub.cs:
```diff
@@ -0,0 +1,72 @@
+            {
+                if (user.Login == userLoginOrEmail || user.Email==userLoginOrEmail)
+                {
+                    if (user.PasswordHash == userPassword) return true;//Тут наверное надо как расшифровывать
```
Вызови просто метод Decrypt отдав ему PasswordHash с БД

DmitriyYakubovskiy committed:
Внес изменения по замечаниям

DmitriyYakubovskiy committed:
.

thisfozz reviewed (dismissed):
Отписал в дм дискорда насчет метода LoginVerification

CozyWool committed:
Добавил автомаппер и модели, и несколько методов по получению данных

thisfozz commented:
![](https://github.trello.services/images/mini-trello-icon.png) [QF. Authorization User (feature-23)](https://trello.com/c/OK6TpUtz/39-qf-authorization-user-feature-23)

thisfozz committed:
Добавил Decrypt пароля при авторизации пользователя

CozyWool committed:
Добавил todo

lexeg commented on Quizzify/Quizzify.Client.Host/ClientHostHub.cs:
```diff
@@ -0,0 +1,116 @@
+using AutoMapper;
```
Я бы предложил рассмотреть такой вариант с авторизацией/регистрацией пользователя. Создать контроллер с двумя POST эндпоинтами. Один для входа пользователя, второй для регистрации.
![image](https://github.com/CozyWool/Quizzify/assets/4533553/0ce25cb8-7db0-4f73-97f1-c32dd1ba0d22)

Как написано про SignalR: 
> SignalR is a free and open-source software library for Microsoft ASP.NET that allows server code to send asynchronous notifications to client-side web applications. 

То есть это библиотека для оповещения других клиентов о чем-либо. Как я понимаю, при аутентификации или регистрации пользователя в вашем проекте оповещать никого не надо. Пользователь только производит операцию входа и еще не вступает в игру. А вот когда пользователь у вас по выданному коду будет входить в одну из викторин, вот тогда должен вступать SignalR в действие.
Подумайте над этим командно.

lexeg commented on Quizzify/Quizzify.Client.Host/ClientHostHub.cs:
```diff
@@ -0,0 +1,116 @@
+
+    public ClientHostHub()
+    {
+        var config = new MapperConfiguration(cfg => cfg.AddProfile<MappingPlayer>());
```
Автомапперы, классы с бизнес-логикой, репозитории и пр. нет необходимости инициализировать в конструкторе, если есть DI-контейнер. Это можно поручить спокойно ему. Например, так:
![image](https://github.com/CozyWool/Quizzify/assets/4533553/77b3c7ba-075f-4685-84db-df90490745a4)

Можно подглядывать сюда: https://github.com/lexeg/ToDoService/tree/main/ToDoService.

И вот еще патч с небольшим примером кода: [0001-.patch](https://github.com/CozyWool/Quizzify/files/14810660/0001-.patch)

DmitriyYakubovskiy committed:
Добавил подключение к игре

добавил эвенты к подключениям к серверам
сделал пару методов для ClientHost
сделал подключение к игре
надо посмотреть переход из авторизейшен модел в плайер модел

thisfozz committed:
Получение выбраннанного вопроса. Добавил мапер для моделей с "Темой"

DmitriyYakubovskiy committed:
Добавил методы для приема и отправки сообщений об ошибке

CozyWool committed:
Вынес Security и Command в отдельные проекты.

CozyWool committed:
Мелкие правки

CozyWool reviewed (approved):
временный слив ветки в мейн, потом доделать необходимо
